implementation-details.md

```diff
@@ -284,32 +313,53 @@ The `CrdsFilter`s are partitioned in pull requests between peers selected random
-> _Example_
->
-> If `mask_bits = 4`, `shard_bits = 12` and `mask = 0001` we search for all shards which contain hashes starting with `0001`, e.g. `000100000000...`, `000100000001...`, and so on.
+As was already [explained](#crds), the `CrdsShards` structure holds a `shards` vector whose elements are an index map of `CrdsValue`s indexes in the `crds` table and their hashes. The `shards` vector allows the node to quickly find and filter out `CrdsValue`s that the pull request sender already has in its `crds`. Let's assume `shard_bits = 3` (actually it is 12), `mask_bits = 3` and `mask = 001 = 1` for simplicity. The `shards` vector would look like this:
 
-3. If `mask_bits > shard_bits` an index whose all bits match the first `shard_bits` of the `mask` is selected, and then values from that index are filtered such that the first `mask_bits` bits of their hashes are equal to the `mask`. Finally, values are filtered using the bloom filter and collected. 
 
-> _Example_
->
-> `mask_bits = 15`, `shard_bits = 12`, `mask = 000001000000001` - first we search for a shard with an index equal to the first 12 bits of the mask, `000001000000`, which is 64, and then for hashes starting with `000001000000001...` from that shard.
+```mermaid
+block-beta
+  block
+    columns 2
+    block
+        columns 1
+        b["000...0"]
+        c["000...1"]
+        d["..."]
+    end
+    l1<["index = 000"]>(left)
+    block
+        columns 1
+        e["001...0"]
+        f["001...1"]
+        g["..."]
+    end
+    l2<["index = 001"]>(left)
+  end
+```
+Numbers in the left column represent the hashes (in binary form) of the `CrdsValue`s stored in the `shards` vector. Elements stored at index `000` contain all hashes starting from `000...` (and corresponding indexes of the `CrdsValue`s in the `crds` table), index `001` contains all hashes starting from `001...` and so on. 
+
+The `mask` value from the pull request `CrdsFilter` is equal to `001` - that means we need to get all elements from the `shards` vector located at index `001`. Since every element contains both hash and the index of the `CrdsValue` in the `crds` table we can quickly find the actual `CrdsValue`s in `crds`. 
+
+When all `CrdsValue`s are collected they are filtered using the provided bloom filter. Elements that don't exist in the filter will be sent back to the pull request sender.
+
+However, there might be the case when `mask_bits < shard_bits` or `mask_bits > shard_bits`:
+
+* `mask_bits < shard_bits` - if e.g. `mask_bits = 2`, `shard_bits = 3` and `mask = 01` we look up all `shards` indexes starting with `01`, e.g. `010`, `011`, `012` and so on, and collect and filter all their elements 
+
+* `mask_bits > shard_bits` - if `mask_bits = 5`, `shard_bits = 3`, `mask = 00101` - first we search for a `shards` index equal to the first 3 bits of the mask, `001`, which is 1, and then for hashes starting with `00101...` from that index.
```

Comment:
any chance you could do a diagram like above for each of these? May be helpful since this algorithm is kinda confusing